config.go

```diff
@@ -0,0 +1,326 @@
+}
+
+func (y *YAML) populate(path []string, i interface{}) error {
+	val, ok := y.at(path)
+	if !ok {
+		return nil
+	}
+	buf := &bytes.Buffer{}
+	if err := yaml.NewEncoder(buf).Encode(val); err != nil {
+		// Provider contents were produced by unmarshaling YAML, this isn't
+		// possible.
+		err := fmt.Errorf(
+			"couldn't marshal config at key %s to YAML: %v",
+			strings.Join(path, _separator),
+			err,
+		)
+		return unreachable.Wrap(err)
+	}
+	dec := yaml.NewDecoder(buf)
+	dec.SetStrict(y.strict)
+	return dec.Decode(i)
+}
+
+func (y *YAML) withDefault(d interface{}) (*YAML, error) {
+	rawDefault := &bytes.Buffer{}
+	if err := yaml.NewEncoder(rawDefault).Encode(d); err != nil {
+		return nil, fmt.Errorf("can't marshal default to YAML: %v", err)
+	}
+	sources := []io.Reader{rawDefault, bytes.NewBuffer(y.raw)}
+	merged, err := merge.YAML(sources, y.strict)
+	if err != nil {
+		// The defaults and the existing config don't agree on the type of a value
+		// (e.g., we're defaulting a mapping to a sequence, which doesn't make
+		// sense).
+		return nil, fmt.Errorf("merging default and existing YAML failed: %v", err)
+	}
+	rawMerged := merged.Bytes()
+
+	var contents interface{}
+	dec := yaml.NewDecoder(merged)
+	dec.SetStrict(y.strict)
+	if err := dec.Decode(&contents); err != nil {
+		return nil, fmt.Errorf("unmarshaling merged YAML failed: %v", err)
+	}
+	return &YAML{
+		name:     y.name,
+		raw:      rawMerged,
+		contents: contents,
+		strict:   y.strict,
+	}, nil
+}
+
+// A Value is a subset of a provider's configuration.
+type Value struct {
+	path     []string
+	provider *YAML
+}
+
+// NewValue is a highly error-prone constructor preserved only for backward
+// compatibility. If value and found don't match the contents of the provider
+// at the supplied key, it panics.
+//
+// Deprecated: this internal constructor was mistakenly exported in the
+// initial release of this package, but its behavior was often very
+// surprising. To guarantee sane behavior without changing the function
+// signature, input validation and panics were added in version 1.2. In all
+// cases, it's both safer and less verbose to use Provider.Get directly.
+func NewValue(p Provider, key string, value interface{}, found bool) Value {
+	actual := p.Get(key)
+	if has := actual.HasValue(); has != found {
+		var tmpl string
+		if has {
+			tmpl = "inconsistent parameters: provider %s has value at key %q but found parameter was false"
+		} else {
+			tmpl = "inconsistent parameters: provider %s has no value at key %q but found parameter was true"
+		}
+		panic(fmt.Sprintf(tmpl, p.Name(), key))
+	}
+	contents := actual.Value()
+	same, err := areSameYAML(contents, value)
+	if err != nil {
+		panic(fmt.Sprintf("can't check NewValue parameter consistency: %v", err))
+	}
+	if !same {
+		tmpl := "inconsistent parameters: provider %s has %#v at key %q but value was %#v"
+		panic(fmt.Sprintf(tmpl, p.Name(), contents, key, value))
+	}
+	return p.Get(key)
+}
```

Comment:
@prashantv and @amckinney: as proposed in #81, here's the new implementation of `NewValue`. Behavior of the returned value is now always the same as `Provider.Get`, at the cost of a panic when the user-supplied parameters are internally inconsistent.